File: apps/desktop/src/atualizacao.ts
/**
 * O que dizer embaixo do botão de atualizar.
 *
 * # Por que isto é um módulo, e não três ternários dentro do painel
 *
 * Porque a pergunta "estou atualizado?" tem **cinco** respostas, e quatro delas
 * eram invisíveis. O painel antigo mostrava recado só nos segundos seguintes ao
 * clique, e tratava dois estados muito diferentes como um só nada:
 *
 * - *conferi e você está em dia* — o caso comum, e o que a pessoa quer ver;
 * - *não consegui conferir* — rede fora, GitHub fora, release sem `latest.json`.
 *
 * Com a mesma cara para os dois, um M/OS que nunca conseguiu falar com o
 * servidor parecia um M/OS atualizado. É daí que sai a impressão de que a
 * atualização "às vezes não funciona": ela às vezes não ACONTECIA, e nunca
 * dizia isso.
 *
 * Separado, o mapa de estado para frase se confere sem abrir o app — e cada
 * estado tem um teste com a data fixa que o produz.
 */

/** O que o Rust grava e devolve. Fatos, nenhuma frase — ver `atualizacao.rs`. */
export type EstadoDaAtualizacao = {
  /** A versão que está rodando agora. */
  versao: string;
  /** Quando esta versão chegou NESTE computador. RFC 3339, ou vazio. */
  instaladaEm: string;
  /** A última verificação que deu certo. Vazio significa nunca. */
  verificadaEm: string;
  /** A versão nova que ela encontrou. Vazio significa que não havia. */
  disponivel: string;
  /** Quando essa versão foi publicada. */
  publicadaEm: string;
  /** O motivo da última tentativa, quando a mais recente falhou. */
  falha: string;
  falhaEm: string;
  endpoint: string;
};

/**
 * As cinco respostas possíveis.
 *
 * `trabalhando` cobre verificar e instalar: nos dois a resposta honesta é "ainda
 * não sei", e inventar um estado separado para cada um dobraria o mapa sem
 * mudar uma palavra da tela.
 */
export type Situacao = "em-dia" | "atrasado" | "sem-resposta" | "nunca" | "trabalhando";

/**
 * Qual das cinco, e **a ordem das perguntas é a decisão**.
 *
 * `disponivel` vem antes de `falha` de propósito: se sabemos que existe uma
 * versão nova, isso continua sendo verdade mesmo que a tentativa de hoje tenha
 * caído — e é a informação sobre a qual dá para agir. Dizer "não consegui
 * verificar" escondendo uma atualização já conhecida seria trocar um fato por
 * uma queixa.
 */
export function situacao(estado: EstadoDaAtualizacao | null, ocupado: boolean): Situacao {
  if (ocupado) return "trabalhando";
  if (!estado) return "nunca";
  // Igual à instalada quer dizer que ela já foi instalada desde a verificação —
  // a anotação é que ficou velha, e não o M/OS.
  if (estado.disponivel && estado.disponivel !== estado.versao) return "atrasado";
  if (estado.falha) return "sem-resposta";
  if (estado.verificadaEm) return "em-dia";
  return "nunca";
}

/** O selo. Curto porque ele é lido de relance, e não lido. */
export function rotulo(situacao: Situacao): string {
  switch (situacao) {
    case "em-dia":
      return "EM DIA";
    case "atrasado":
      return "DESATUALIZADO";
    case "sem-resposta":
      return "NÃO CONFERIDO";
    case "nunca":
      return "NUNCA CONFERI";
    case "trabalhando":
      return "CONFERINDO";
  }
}

/** "26/08, 02:31". Sem ano: a versão que roda é sempre recente. */
function dia(iso: string): string {
  const quando = new Date(iso);
  if (Number.isNaN(quando.getTime())) return "";
  return quando.toLocaleString("pt-BR", {
    day: "2-digit",
    month: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
  });
}

/** Só o dia, para a data de publicação — a hora de um release não decide nada. */
function data(iso: string): string {
  const quando = new Date(iso);
  if (Number.isNaN(quando.getTime())) return "";
  return quando.toLocaleDateString("pt-BR", { day: "2-digit", month: "2-digit" });
}

/**
 * A primeira linha: o que está rodando, e desde quando.
 *
 * A data vem do carimbo do executável, então ela responde "desde quando este
 * computador está nesta versão" — que é a pergunta —, e não "quando isto foi
 * compilado".
 */
export function linhaDaVersao(estado: EstadoDaAtualizacao | null): string {
  if (!estado?.versao) return "Versão —";
  const quando = dia(estado.instaladaEm);
  return quando ? `Versão ${estado.versao} · instalada em ${quando}` : `Versão ${estado.versao}`;
}

/**
 * A segunda linha: o que a última verificação descobriu.
 *
 * `relativa` é injetada em vez de importada para o teste poder fixar o tempo. O
 * painel passa a `relativeTime` de sempre.
 */
export function linhaDaVerificacao(
  estado: EstadoDaAtualizacao | null,
  relativa: (iso: string) => string,
): string {
  const qual = situacao(estado, false);
  if (!estado) return "";

  switch (qual) {
    case "atrasado": {
      const publicada = data(estado.publicadaEm);
      return publicada
        ? `Há uma versão nova: ${estado.disponivel}, publicada em ${publicada}.`
        : `Há uma versão nova: ${estado.disponivel}.`;
    }
    case "em-dia":
      return `Conferido ${relativa(estado.verificadaEm)}. Nenhuma versão nova.`;
    case "sem-resposta": {
      // As duas metades, e não uma: o que se soube antes continua valendo, e
      // apagá-lo transformaria uma queda de rede em "você nunca verificou".
      const queixa = `Não consegui conferir ${relativa(estado.falhaEm)}: ${estado.falha}`;
      return estado.verificadaEm
        ? `${queixa} A última vez que deu certo foi ${relativa(estado.verificadaEm)}.`
        : queixa;
    }
    case "nunca":
      return "Ainda não conferi se existe versão nova.";
    case "trabalhando":
      return "";
  }
}

/**
 * Já passou tempo bastante para conferir sozinho?
 *
 * O M/OS confere na abertura para o selo significar alguma coisa: um indicador
 * que só se atualiza quando alguém entra em Settings e clica é um indicador que
 * mostra o passado. Mas conferir a cada abertura seria uma ida à rede toda vez
 * que a janela nasce — e o M/OS abre no logon.
 *
 * Seis horas: mais de uma vez por dia de trabalho, e longe de uma por abertura.
 * Uma falha NÃO reinicia a contagem por si só; ela conta como tentativa, senão
 * um GitHub fora do ar viraria uma tentativa por minuto.
 */
export function deveConferirSozinho(
  estado: EstadoDaAtualizacao | null,
  agora: Date = new Date(),
  horas = 6,
): boolean {
  if (!estado) return false;
  // Já sabemos que há uma nova: não há o que descobrir, e a tela já diz.
  if (estado.disponivel && estado.disponivel !== estado.versao) return false;

  const ultima = [estado.verificadaEm, estado.falhaEm]
    .map((iso) => new Date(iso).getTime())
    .filter((instante) => !Number.isNaN(instante));
  if (!ultima.length) return true;

  return agora.getTime() - Math.max(...ultima) > horas * 3600_000;
}

File: apps/desktop/src-tauri/src/atualizacao.rs
//! O que a tela precisa para responder "estou atualizado?".
//!
//! # Por que este modulo existe
//!
//! O painel de atualizacoes sabia responder a pergunta **so durante os segundos
//! seguintes ao clique**. Verificar, sair de Settings e voltar apagava tudo: o
//! estado morava em `useState`, e desmontar o componente levava junto a unica
//! prova de que a verificacao tinha acontecido. Na pratica isso queria dizer que
//! o M/OS nunca sabia dizer se estava em dia — so sabia dizer se voce tinha
//! clicado ha pouco.
//!
//! Pior: **"nao havia versao nova" e "nao consegui verificar" tinham a mesma
//! cara** — nenhuma. Uma verificacao que falhou por falta de rede deixava a tela
//! igualzinha a uma que deu certo, e e exatamente dai que sai o "as vezes ele
//! nao funciona": o app parecia ter conferido quando nao tinha.
//!
//! Entao o resultado da verificacao passa a ser um FATO GRAVADO, e nao estado de
//! tela. Ele mora no `settings.json`, ao lado das outras preferencias deste
//! aparelho, porque e disso que se trata: uma anotacao local sobre esta
//! instalacao. Ele nao sincroniza — a versao instalada e um fato do aparelho, e
//! mandar isso para o hub faria o celular achar que tambem esta desatualizado.
//!
//! # O dia da versao que esta rodando
//!
//! Vem do carimbo do proprio executavel no disco. Isso responde **de quando e a
//! versao que esta rodando**, e nao "desde quando este computador esta nela": o
//! NSIS PRESERVA o carimbo dos arquivos que empacota, entao o `mos-desktop.exe`
//! mantem a hora em que o CI o compilou, e nao a hora em que o instalador
//! passou por aqui. Conferido em 28/08/2026: o executavel marcava 25/08 23:30,
//! que e o build do release; o `uninstall.exe`, que o NSIS gera na hora, marcava
//! 26/08 09:42, que e a instalacao.
//!
//! A pergunta que a tela faz e a primeira — "de que dia e a atualizacao que ele
//! esta executando" —, entao o carimbo serve. O nome do campo diz isso, e nao
//! "instalada em": um rotulo que promete a data da instalacao e entrega a da
//! compilacao e uma mentira pequena que ninguem tem como perceber.

use std::path::Path;

use serde::Serialize;
use time::format_description::well_known::Rfc3339;
use time::OffsetDateTime;

use crate::{load_settings, save_settings, AppState};

/// O que a tela desenha embaixo do botao.
///
/// Tudo aqui e fato, e nenhuma frase: quem escreve a frase e a interface, que e
/// quem conhece o idioma e o espaco disponivel. Um campo `mensagem` pronto aqui
/// obrigaria o Rust a decidir o texto de uma tela que ele nao ve.
#[derive(Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EstadoDaAtualizacao {
    /// A versao que esta rodando agora.
    pub versao: String,
    /// De quando E esta versao, em RFC 3339: o carimbo do executavel, que o
    /// instalador preserva do build. Vazio quando ele nao pode ser lido — que
    /// acontece, e nao e erro. Ver o cabecalho para por que nao e "instalada em".
    pub versao_de: String,
    /// Quando a ultima verificacao BEM-SUCEDIDA aconteceu. Vazio significa
    /// nunca, e "nunca" e uma resposta que a tela precisa poder dar.
    pub verificada_em: String,
    /// A versao nova que a ultima verificacao encontrou. Vazio significa que ela
    /// encontrou o que se espera encontrar: nada.
    pub disponivel: String,
    /// Quando a versao disponivel foi publicada, em RFC 3339.
    pub publicada_em: String,
    /// O motivo da ultima verificacao que FALHOU, quando a mais recente falhou.
    ///
    /// Separado de `verificada_em` de proposito: as duas coexistem, e a tela
    /// precisa das duas para dizer "voce estava em dia ontem, e hoje eu nao
    /// consegui conferir" em vez de escolher uma das metades.
    pub falha: String,
    pub falha_em: String,
    /// De onde a verificacao busca. Aparece na tela porque, quando ela falha, a
    /// primeira pergunta util e "falha ao falar com quem?".
    pub endpoint: String,
}

/// De onde o `tauri.conf.json` manda buscar.
///
/// Repetido aqui como texto, e a repeticao e deliberada: ler o JSON de
/// configuracao em tempo de execucao so para mostrar uma linha na tela custaria
/// um caminho de arquivo e um parse que podem falhar, para exibir algo que nao
/// muda entre versoes. Se o endpoint mudar, esta linha muda junto — e o teste
/// que compara os dois nao existe porque o custo do erro e uma URL errada num
/// texto de diagnostico, e nao uma atualizacao que nao chega.
const ENDPOINT: &str = "https://github.com/theusinshow/m-os/releases/latest/download/latest.json";

/// Quando o executavel que esta rodando foi escrito no disco.
///
/// Na pratica, a hora em que o CI o compilou: o NSIS preserva o carimbo do que
/// empacota. Ver o cabecalho.
///
/// Falha em silencio — devolvendo vazio — porque nao saber a data e uma
/// informacao a menos, e nao um motivo para o painel inteiro nao abrir.
fn carimbo_do_executavel() -> String {
    std::env::current_exe()
        .ok()
        .and_then(|caminho| carimbo_de(&caminho))
        .unwrap_or_default()
}

fn carimbo_de(caminho: &Path) -> Option<String> {
    let modificado = std::fs::metadata(caminho).ok()?.modified().ok()?;
    OffsetDateTime::from(modificado).format(&Rfc3339).ok()
}

#[tauri::command]
pub fn atualizacao_estado(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
) -> EstadoDaAtualizacao {
    let settings = load_settings(&state.settings_path);
    EstadoDaAtualizacao {
        versao: app.package_info().version.to_string(),
        versao_de: carimbo_do_executavel(),
        verificada_em: settings.atualizacao_verificada_em.clone(),
        disponivel: settings.atualizacao_disponivel.clone(),
        publicada_em: settings.atualizacao_publicada_em.clone(),
        falha: settings.atualizacao_falha.clone(),
        falha_em: settings.atualizacao_falha_em.clone(),
        endpoint: ENDPOINT.to_owned(),
    }
}

/// Anota uma verificacao que DEU CERTO.
///
/// `disponivel` vazio significa "conferi e nao ha versao nova" — que e o
/// resultado comum, e o unico que a tela antiga nao conseguia guardar.
///
/// Apagar a falha aqui e o ponto: uma verificacao bem-sucedida torna a queixa
/// anterior historia, e deixar as duas na tela faria o painel contar duas
/// versoes da mesma coisa.
#[tauri::command]
pub fn atualizacao_anotar_verificacao(
    state: tauri::State<'_, AppState>,
    disponivel: String,
    publicada_em: String,
) -> Result<(), mos_core::CoreError> {
    let mut settings = load_settings(&state.settings_path);
    settings.atualizacao_verificada_em = agora();
    settings.atualizacao_disponivel = disponivel;
    settings.atualizacao_publicada_em = publicada_em;
    settings.atualizacao_falha.clear();
    settings.atualizacao_falha_em.clear();
    save_settings(&state.settings_path, &settings)
}

/// Anota uma verificacao que FALHOU.
///
/// Ela NAO apaga `verificada_em` nem `disponivel`: o que se soube ontem continua
/// sendo o que se sabe. Apagar transformaria uma queda de rede em "voce nunca
/// verificou", que e mentira e e pior — some a unica informacao que ainda valia.
#[tauri::command]
pub fn atualizacao_anotar_falha(
    state: tauri::State<'_, AppState>,
    motivo: String,
) -> Result<(), mos_core::CoreError> {
    let mut settings = load_settings(&state.settings_path);
    settings.atualizacao_falha = motivo;
    settings.atualizacao_falha_em = agora();
    save_settings(&state.settings_path, &settings)
}

fn agora() -> String {
    OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_default()
}

#[cfg(test)]
mod testes {
    use super::*;

    #[test]
    fn o_carimbo_de_um_arquivo_que_existe_e_legivel_como_rfc3339() {
        let pasta = tempfile::tempdir().unwrap();
        let arquivo = pasta.path().join("mos-desktop.exe");
        std::fs::write(&arquivo, b"binario de mentira").unwrap();

        let carimbo = carimbo_de(&arquivo).expect("um arquivo recem-escrito tem carimbo");
        // A tela formata isto com `new Date(...)`, entao ele precisa ser um
        // instante que o JavaScript entenda — nao basta ser uma string.
        OffsetDateTime::parse(&carimbo, &Rfc3339).expect("o carimbo precisa ser RFC 3339");
    }

    #[test]
    fn o_carimbo_de_um_arquivo_ausente_e_vazio_e_nao_erro() {
        // Nao saber a data e uma informacao a menos, e nao motivo para o painel
        // inteiro nao abrir.
        assert!(carimbo_de(Path::new("nao/existe/mos-desktop.exe")).is_none());
    }
}
